SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -86,53 +120,78 @@ TArray<SpatialLoadBalanceEnforcer::AclWriteAuthorityRequest> SpatialLoadBalanceE
-			CompletedRequests.Add(Request.EntityId);
+			UE_LOG(LogSpatialLoadBalanceEnforcer, Warning, TEXT("Entity with invalid virtual worker ID assignment will not be processed. EntityId: %lld. This should not happen - investigate if you see this warning."), EntityId);
+			CompletedRequests.Add(EntityId);
 			continue;
 		}
 
 		check(VirtualWorkerTranslator != nullptr);
 		const PhysicalWorkerName* DestinationWorkerId = VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
 		if (DestinationWorkerId == nullptr)
 		{
-			const int32 WarnOnAttemptNum = 5;
-			Request.ProcessAttempts++;
-			// TODO(zoning): Revisit this when we support replacing crashed workers.
-			if (Request.ProcessAttempts % WarnOnAttemptNum == 0)
-			{
-				UE_LOG(LogSpatialLoadBalanceEnforcer, Warning, TEXT("Failed to process WriteAuthAssignmentRequest because virtual worker mapping is unset for EntityID: %lld. Process attempts made: %d"), Request.EntityId, Request.ProcessAttempts);
-			}
-			// A virtual worker -> physical worker mapping may not be established yet.
-			// We'll retry on the next Tick().
+			UE_LOG(LogSpatialLoadBalanceEnforcer, Error, TEXT("This worker is not assigned a virtual worker. This shouldn't happen! Worker: %s"), *WorkerId);
 			continue;
 		}
 
-		if (StaticComponentView->HasAuthority(Request.EntityId, SpatialConstants::ENTITY_ACL_COMPONENT_ID))
+		if (StaticComponentView->HasAuthority(EntityId, SpatialConstants::ENTITY_ACL_COMPONENT_ID))
 		{
-			PendingRequests.Push(AclWriteAuthorityRequest{ Request.EntityId, *DestinationWorkerId });
+			EntityAcl* Acl = StaticComponentView->GetComponentData<EntityAcl>(EntityId);
+			// We always assign a client endpoint requirement set. It might be empty, but to be consistent with the previous state we grab it here anyway.
+			WorkerRequirementSet* ClientRequirementSet = Acl->ComponentWriteAcl.Find(SpatialConstants::GetClientAuthorityComponent(GetDefault<USpatialGDKSettings>()->UseRPCRingBuffer()));
+			TArray<Worker_ComponentId> ComponentIds;
+			Acl->ComponentWriteAcl.GetKeys(ComponentIds);
+			PendingRequests.Push(
+				AclWriteAuthorityRequest{
+					EntityId,
+					*DestinationWorkerId,
+					Acl->ReadAcl,
+					*ClientRequirementSet,
+					ComponentIds
+				});
 		}
 		else
 		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Failed to update the EntityACL to match the authority intent; this worker does not have authority over the EntityACL."
-				" Source worker ID: %s. Entity ID %lld. Desination worker ID: %s."), *WorkerId, Request.EntityId, **DestinationWorkerId);
+			UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Failed to update the EntityACL to match the authority intent; this worker does lost authority over the EntityACL since the request was queued."
```

Comment:
```suggestion
			UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Failed to update the EntityACL to match the authority intent; this worker lost authority over the EntityACL since the request was queued."
```

Reply:
thanks, good spot